### ISSUE TITLE
Scheduled state not able to send output to returner. 

### ISSUE THREAD

**Issue by rbthomp** [saltstack/salt#49160]
### Description of Issue/Question
I'm getting an error when trying to schedule a state run, with a returner. 

The data never gets sent to the returner rawfile_json.

```
2018-08-16 10:09:42,951 [salt.utils.process:754 ][ERROR   ][31361] An un-handled exception from the multiprocessing process 'SignalHandlingMultiprocessingProcess-1:6' was caught:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/utils/process.py", line 747, in _run
    return self._original_run()
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python2.7/site-packages/salt/utils/schedule.py", line 757, in handle_func
    if not data_returner and not self.schedule_returner:
UnboundLocalError: local variable 'data_returner' referenced before assignment
```
### Setup
Schedule a state run using pillars. 
/etc/salt/minion.d/salt_events.conf
```
rawfile_json.filename: /var/log/salt/events
```
pillar
```
schedule:
      test_schedule:
          function: state.sls
          returner: rawfile_json
          minutes: 2
          args:
            - utils.test
          splay: 1
````

utils/test.sls
```
always-passes:
  test.succeed_without_changes:
    - name: foo
```

### Steps to Reproduce Issue
Configure schedule to run via pillars.

### Versions Report
Salt Version:
           Salt: 2018.3.0

Dependency Versions:
           cffi: 1.5.2
       cherrypy: 3.6.0
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: 0.24.0
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.24.0
         Python: 2.7.13 (default, Jan 11 2017, 10:56:06) [GCC]
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 14.0.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.4

System Versions:
           dist: SuSE 12 x86_64
         locale: UTF-8
        machine: x86_64
        release: 4.4.140-94.42-default
         system: Linux
        version: SUSE Linux Enterprise Server  12 x86_64

**Comment by garethgreenaway** [comment]
@rbthomp So far I haven't been able to reproduce this particular error.  I'm wondering if there is another exception that is happening before this one and causing the `data_returner` variable to not be available i when it's referenced later.

**Comment by garethgreenaway** [comment]
@rbthomp I was finally able to reproduce this under the following scenario, a Salt master running 2017.74 and a Salt minion running 2018.3.  Does this match your setup at all?

**Comment by rbthomp** [comment]
Thanks for your help on this. You were correct there is an error before the one I provide. Here is the full error result.
**ERROR**
```
2018-08-17 08:26:29,169 [salt.utils.schedule:738 ][ERROR   ][1276] Unhandled exception running state.sls
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/utils/schedule.py", line 705, in handle_func
    ret['return'] = self.functions[func](*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/modules/state.py", line 1296, in sls
    _set_retcode(ret, high_)
  File "/usr/lib/python2.7/site-packages/salt/modules/state.py", line 106, in _set_retcode
    if not __utils__['state.check_result'](ret, highstate=highstate):
  File "/usr/lib/python2.7/site-packages/salt/utils/context.py", line 228, in __getitem__
    return self._dict()[key]
KeyError: u'state.check_result'
2018-08-17 08:26:29,177 [salt.utils.process:754 ][ERROR   ][1276] An un-handled exception from the multiprocessing process 'SignalHandlingMultiprocessingProcess-1:17' was caught:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/utils/process.py", line 747, in _run
    return self._original_run()
  File "/usr/lib64/python2.7/multiprocessing/process.py", line 114, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python2.7/site-packages/salt/utils/schedule.py", line 757, in handle_func
    if not data_returner and not self.schedule_returner:
UnboundLocalError: local variable 'data_returner' referenced before assignment
```

Here is the version for the master and minion.

**Master:**
Salt Version:
           Salt: 2018.3.0

Dependency Versions:
           cffi: 1.5.2
       cherrypy: 3.6.0
       dateutil: 2.4.2
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.8
        libgit2: 0.24.0
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: 2.10
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.24.0
         Python: 2.7.13 (default, Jan 11 2017, 10:56:06) [GCC]
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 14.0.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.4

System Versions:
           dist: SuSE 12 x86_64
         locale: UTF-8
        machine: x86_64
        release: 4.4.140-94.42-default
         system: Linux
        version: SUSE Linux Enterprise Server  12 x86_64

**Minion:**
Salt Version:
           Salt: 2018.3.0

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.21.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Jul 13 2018, 13:06:57)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.5.0
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5

System Versions:
           dist: centos 7.5.1804 Core
         locale: UTF-8
        machine: x86_64
        release: 3.10.0-862.9.1.el7.x86_64
         system: Linux
        version: CentOS Linux 7.5.1804 Core

**Comment by rbthomp** [comment]
@garethgreenaway I have an update on this it looks like the loader __utils__['state.check_result'] is being called in salt/modules/state.py, but for whatever reason the  state.check_result is not in the __utils__ dict. Don't understand why, I don't know exactly how the loader work. The thing is salt.utils.state is included in the state module, so simply calling the function using salt.utils.state.check_result instead of __utils__['state.check_result'] corrected the issue. I was then able to get the returner working correctly. 

What are your thoughts on this?

**Comment by garethgreenaway** [comment]
@rbthomp Apologies for the delay on this one.  The scenario you describe above makes sense if you're running an older Salt master and a newer Salt minion, the minion is attempting to send a command back to the Salt master which is unavailable because it's the older version and wasn't implemented.  Officially we do not support these scenarios where the Salt master is older than the Salt minions for these reasons.

**Comment by rbthomp** [comment]
@garethgreenaway I've done additional testing. Here are my result.

No Errors, Works Fine.
```
CentOS-7:
Salt-Master = 2018.3.2

CentOS-7:
Salt-Minion = 2018.3.2 

No Errors.
```

Errors no returned data.
```
CentOS-7:
Salt-Master = 2018.3.2

SLES-12:
Salt-Minion = 2018.3.0 
```
I'm fine with just updating all my minions to 2018.3.2 to eliminate the issue and I have on CentOS. My issue is SLES-12 hasn't released 2018.3.2 through their channels and no SaltStack maintained repo exists for SLES-12.

**Comment by OrangeDog** [comment]
I'm getting similar error with both master and minion on 2018.3.3, apparently after a timeout.
```
2019-01-14 14:48:08,629 [salt.utils.schedule:740 ][ERROR   ][3869] Unhandled exception running state.apply
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/schedule.py", line 707, in handle_func
    ret['return'] = self.functions[func](*args, **kwargs)
  File "/usr/lib/python3/dist-packages/salt/modules/state.py", line 764, in apply_
    return highstate(**kwargs)
  File "/usr/lib/python3/dist-packages/salt/modules/state.py", line 1044, in highstate
    initial_pillar=_get_initial_pillar(opts))
  File "/usr/lib/python3/dist-packages/salt/state.py", line 4046, in __init__
    BaseHighState.__init__(self, opts)
  File "/usr/lib/python3/dist-packages/salt/state.py", line 2912, in __init__
    self.avail = self.__gather_avail()
  File "/usr/lib/python3/dist-packages/salt/state.py", line 2922, in __gather_avail
    avail[saltenv] = self.client.list_states(saltenv)
  File "/usr/lib/python3/dist-packages/salt/fileclient.py", line 392, in list_states
    for path in self.file_list(saltenv):
  File "/usr/lib/python3/dist-packages/salt/fileclient.py", line 1284, in file_list
    else self.channel.send(load)
  File "/usr/lib/python3/dist-packages/salt/utils/asynchronous.py", line 64, in wrap
    ret = self._block_future(ret)
  File "/usr/lib/python3/dist-packages/salt/utils/asynchronous.py", line 74, in _block_future
    return future.result()
  File "/usr/lib/python3/dist-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib/python3/dist-packages/salt/transport/zeromq.py", line 325, in send
    ret = yield self._crypted_transfer(load, tries=tries, timeout=timeout, raw=raw)
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/usr/lib/python3/dist-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib/python3/dist-packages/salt/transport/zeromq.py", line 293, in _crypted_transfer
    ret = yield _do_transfer()
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/usr/lib/python3/dist-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 1063, in run
    yielded = self.gen.throw(*exc_info)
  File "/usr/lib/python3/dist-packages/salt/transport/zeromq.py", line 277, in _do_transfer
    tries=tries,
  File "/usr/lib/python3/dist-packages/tornado/gen.py", line 1055, in run
    value = future.result()
  File "/usr/lib/python3/dist-packages/tornado/concurrent.py", line 238, in result
    raise_exc_info(self._exc_info)
  File "<string>", line 4, in raise_exc_info
salt.exceptions.SaltReqTimeoutError: Message timed out
2019-01-14 14:48:08,715 [salt.utils.process:754 ][ERROR   ][3869] An un-handled exception from the multiprocessing process 'SignalHandlingMultiprocessingProcess-1:499' was caught:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/salt/utils/process.py", line 747, in _run
    return self._original_run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/usr/lib/python3/dist-packages/salt/utils/schedule.py", line 759, in handle_func
    if not data_returner and not self.schedule_returner:
UnboundLocalError: local variable 'data_returner' referenced before assignment
```

**Comment by crux-capacitor** [comment]
Got the same error with both master and minion running 2018.3.3.

**Comment by garethgreenaway** [comment]
Unfortunately I haven't been able to reproduce this.  If you're experiencing this issue, can you provide the contents of the Salt schedule, eg. results from `/etc/salt/minion.d/_schedule.conf`.  Thanks!

**Comment by garethgreenaway** [comment]
Looking closer at the code and the differences between 2018.3.3 and the HEAD of 2018.3, I see why this particular UnboundLocalError is happening.  Something is causing this try...except to be triggered, https://github.com/saltstack/salt/blob/v2018.3.3/salt/utils/schedule.py#L628, and the data_returner variable is being created inside that try...except.  In the HEAD of 2018.3 that was moved outside the try..except block.  Something above this line, https://github.com/saltstack/salt/blob/v2018.3.3/salt/utils/schedule.py#L716, is causing the exception so when data_returner is referenced in the `finally` section it doesn't exist.  This particular issue will be fixed in the next 2018.3 release and future releases as well.

**Comment by vquiering** [comment]
@garethgreenaway, I couldn't find a PR. Is there anything we can already test?
